src/query/atomWithInfiniteQuery.ts

```diff
@@ -0,0 +1,164 @@
+      let settlePromise:
+        | ((data: InfiniteData<TData> | null, err?: TError) => void)
+        | null = null
+      const getInitialData = () =>
+        typeof options.initialData === 'function'
+          ? (
+              options.initialData as InitialDataFunction<
+                InfiniteData<TQueryData>
+              >
+            )()
+          : options.initialData
+      const dataAtom = atom<
+        | InfiniteData<TData | TQueryData>
+        | Promise<InfiniteData<TData | TQueryData>>
+      >(
+        getInitialData() ||
+          new Promise<InfiniteData<TData>>((resolve, reject) => {
+            settlePromise = (data, err) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(data as InfiniteData<TData>)
+              }
+            }
+          })
+      )
+      let setData: (
+        data: InfiniteData<TData> | Promise<InfiniteData<TData>>
+      ) => void = () => {
+        throw new Error('atomWithInfiniteQuery: setting data without mount')
+      }
+      let prevData: InfiniteData<TData> | null = null
+
+      const listener = (
+        result:
+          | QueryObserverResult<InfiniteData<TData>, TError>
+          | { data?: undefined; error: TError }
+      ) => {
+        if (result.error) {
+          if (settlePromise) {
+            settlePromise(null, result.error)
+            settlePromise = null
+          } else {
+            setData(Promise.reject<InfiniteData<TData>>(result.error))
+          }
+          return
+        }
+        if (
+          result.data === undefined ||
+          (prevData !== null && equalityFn(prevData, result.data))
+        ) {
+          return
+        }
+        prevData = result.data
+        if (settlePromise) {
+          settlePromise(result.data)
+          settlePromise = null
+        } else {
+          setData(result.data)
+        }
+      }
+
+      const defaultedOptions = queryClient.defaultQueryObserverOptions(options)
+
+      if (typeof defaultedOptions.staleTime !== 'number') {
+        defaultedOptions.staleTime = 1000
+      }
+
+      const observer = new InfiniteQueryObserver(queryClient, defaultedOptions)
+
+      observer
+        .fetchOptimistic(defaultedOptions)
+        .then(listener)
+        .catch((error) => listener({ error }))
+
+      dataAtom.onMount = (update) => {
+        setData = update
+        const unsubscribe = observer?.subscribe(listener)
+        return unsubscribe
+      }
+      return { dataAtom, observer, options }
+    },
+    (get, set, action: AtomWithInfiniteQueryAction) => {
+      const { observer } = get(queryDataAtom)
+      switch (action.type) {
+        case 'refetch': {
+          void observer.refetch()
+          break
+        }
+        case 'fetchPreviousPage': {
+          void observer.fetchPreviousPage()
+          break
+        }
+        case 'fetchNextPage': {
+          void observer.fetchNextPage()
```

Comment:
Is this `void` for TS? `fetchNextPage()` returns a promise? We could return `Promise<void>`. Did so in atomWithQuery's refetch.

Reply:
yeah this is to ignore the promise. I tried a version that did the same void'ed promise return but it didn't do anything lol so I think this is fine for now. I also don't know the best route to pass the status via this atom. I have a different atom that can take a query key and return the status. maybe we can discuss it and come back to it?